src/Raven.Server/Smuggler/Documents/MultiShardedDestination.cs

```diff
@@ -242,10 +244,50 @@ public ShardedDocumentActions(ShardedDatabaseContext databaseContext, ByteString
 
             public async ValueTask WriteDocumentAsync(DocumentItem item, SmugglerProgressBase.CountsWithLastEtagAndAttachments progress, Func<ValueTask> beforeFlushing = null)
             {
-                var shardNumber = DatabaseContext.GetShardNumberFor(_allocator, item.Document.Id);
+                var shardNumber = GetShardNumberAndHandleLegacyRevisionsIfNeeded(item.Document);
                 await _actions[shardNumber].WriteDocumentAsync(item, progress, beforeFlushing);
             }
 
+            private int GetShardNumberAndHandleLegacyRevisionsIfNeeded(Document document)
+            {
+                var legacy = document.NonPersistentFlags.Contain(NonPersistentDocumentFlags.LegacyRevision) ||
+                             document.NonPersistentFlags.Contain(NonPersistentDocumentFlags.LegacyHasRevisions);
+                if (legacy == false)
+                    return DatabaseContext.GetShardNumberFor(_allocator, document.Id);
+
+                string id = document.Id;
+                if (document.Data.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata))
+                {
+                    JsonOperationContext context = GetContextForNewDocument();
+                    if (document.NonPersistentFlags.Contain(NonPersistentDocumentFlags.LegacyRevision))
+                    {
+                        var endIndex = document.Id.IndexOf(DatabaseSmuggler.PreV4RevisionsDocumentId, StringComparison.OrdinalIgnoreCase);
+                        id = document.Id.Substring(0, endIndex);
+
+                        metadata.Modifications = new DynamicJsonValue
+                        {
+                            ["Raven-Document-Parent-Revision"] = document.Id,
```

Comment:
we do not do that anywhere else in the code... how this is handled for regular database?

Reply:
In the case of regular DB, we read it directly from the metadata